RT_UPDATE/download_asfr.R

```diff
@@ -0,0 +1,363 @@
+#' asfr_data <- download_Asfr(location = "France",
+#'                            AsfrDatesIn = c(2000, 2005, 2010),
+#'                            output = "5-year")
+#'
+#' # Use user-supplied ASFR matrix (rows = ages, cols = years)
+#' Asfrmat <- matrix(runif(25), nrow = 5)
+#' download_Asfr(Asfrmat = Asfrmat,
+#'               AsfrDatesIn = c(2000, 2005, 2010, 2015, 2020),
+#'               output = "single")
+#' }
+#'
+#' @export
+
+
+download_Asfr <- function(Asfrmat     = NULL,
+                          location    = NULL,
+                          AsfrDatesIn = NULL,
+                          method      = "linear",
+                          output      = "5-year",
+                          ...) {
+  
+  verbose <- getOption("basepop_verbose", TRUE)
+  
+  # -------------------------------
+  # CASE 1: user-supplied Asfrmat
+  # -------------------------------
+  # If the user provides Asfrmat we simply coerce to matrix, assign column names
+  # (years) and set rownames (ages) according to requested output format.
+  #
+  # NOTE: Asfrmat is expected to be in conventional reproductive ages
+  # (usually 15-49 single-year ages or 5-year groups 15-19,...). If input
+  # has a different age basis, it should be transformed first.
+  if(!is.null(Asfrmat)) {
+    
+    Asfrmat           <- as.matrix(Asfrmat)
+    colnames(Asfrmat) <- AsfrDatesIn
+    n                 <- nrow(Asfrmat)
+    
+    # Assign age labels depending on output. Keep behaviour consistent with earlier code.
+    if(output == "single") {
+      
+      # Single-year ages: assume row 1 corresponds to age 10 (user responsibility)
+      Age <- 10:(n - 1)
+      
+    }
+    
+    # 5-year groups
+    if(output == "5-year") { 
+      
+      Age <- seq(from = 10, length.out = n, by = 5)
+      
+    }
+    
+    rownames(Asfrmat) <- Age
+    
+    return(Asfrmat)
+    
+  }
+  
+  # -------------------------------
+  # CASE 2: download from WPP
+  # -------------------------------
+  if(is.null(location)) {
+    
+    stop("You need to provide a location to download the data for Asfrmat")
+    
+  }
+  
+  # ------------------------------------------------------ #
+  # installed wpp versions
+  installed_wpp <- grep("^wpp\\d{4}$", rownames(installed.packages()), value = TRUE)
+  
+  # stop if none
+  if(length(installed_wpp) == 0) { 
+    
+    stop("No wpp package installed.")
+    
+  }
+  # find the lates one
+  latest_wpp <- sort(installed_wpp, decreasing = TRUE)[1]
+  
+  if(latest_wpp < 2022) {
+    
+    stop("No single ages are availabe in wpp versions earlier that wpp2022.
+           Please update the wpp package.")
+    
+  }
+  
+  # download mx1dt data from the latest package version
+  data("percentASFR1dt", package = latest_wpp)
+  data("tfr1",           package = latest_wpp)
+  
+  # if any date chosen is less then 1950 or more than wpp version + 1
+  if(any(AsfrDatesIn < 1950, AsfrDatesIn > (parse_number(latest_wpp) + 1))) {
```

Comment:
check not needed? Can pad w default or avg or something else? if we do this then message() or warning()